### PR TITLE
don't scale when adding to sprite atlas

### LINKED ISSUE #3164: never scale images when copying them into the SpriteAtlas
Currently we bilinearly scale images that are not at the native pixel ratio up or down when adding them to the SpriteAtlas. This has led to code that is more complex than necessary and produced issues like #3031 and #2198.

Instead, we should copy images into the SpriteAtlas pixel-for-pixel, and let up/down sampling happen in the shaders.

@kkaefer @ansis -- am I missing anything that would make this non-viable?

### PR DESCRIPTION
**Pull request by ansis**
Don't scale images when adding them to the sprite atlas and make the sprite atlas match -js.

fix #3164 and port https://github.com/mapbox/mapbox-gl-js/pull/1919/files

can you review @jfirebaugh?

### CODE REVIEW

**Comment by jfirebaugh** [comment]
Also fixes #3513?

**Comment by mb12** [comment]
@ansis Can SpriteAtlas::bind be unconditionally called with linear=true?

**Comment by jfirebaugh** [comment]
What about [this scaling code](https://github.com/mapbox/mapbox-gl-native/blob/sprite-atlas-no-scaling/src/mbgl/sprite/sprite_atlas.cpp#L132-L152)?

**Comment by ansis** [comment]
> Also fixes #3513?

Yes, except for https://github.com/mapbox/mapbox-gl-native/issues/3528

> @ansis Can SpriteAtlas::bind be unconditionally called with linear=true?

Using nearest neighbour filtering is clearer when icons copied at their original size. This small bit of extra clarity matters since icons are already pretty small.

> What about this scaling code?

yeah, that needs to be removed.

**Comment by ansis** [comment]
The nearest neighbour scaling code is removed in fc20199

**Comment by ansis** [comment]
@jfirebaugh I think this should be good now. Want to take another look?

File: src/mbgl/sprite/sprite_atlas.cpp [hunk 1]
```diff
@@ -83,28 +83,48 @@ mapbox::util::optional<SpriteAtlasPosition> SpriteAtlas::getPosition(const std::
...
 
+void copyBitmap(const uint32_t *src, const uint32_t srcStride, const uint32_t srcX, const uint32_t srcY,
+        uint32_t *const dst, const uint32_t dstStride, const uint32_t dstX, const uint32_t dstY, int dstSize,
+        const int width, const int height, const bool wrap) {
+
+    int srcI = srcY * srcStride + srcX;
+    int dstI = dstY * dstStride + dstX;
+    int x, y;
+
+    if (wrap && false) {
```
Comment (jfirebaugh) [inline src/mbgl/sprite/sprite_atlas.cpp]: ?

File: src/mbgl/util/scaling.hpp [hunk 2]
```diff
@@ -10,10 +10,6 @@
 namespace mbgl {
 namespace util {
 
-void bilinearScale(const uint32_t* srcData, const vec2<uint32_t>& srcSize,
```
Comment (jfirebaugh) [inline src/mbgl/util/scaling.hpp]: I believe this whole file can be deleted.

File: src/mbgl/sprite/sprite_atlas.cpp (hunk 1 above)
Comment (ansis) [inline src/mbgl/sprite/sprite_atlas.cpp]: oh, good catch. I'll remove and add a render test to catch this kind of bug: https://github.com/mapbox/mapbox-gl-test-suite/issues/72

**Comment by ansis** [comment]
scaling.hpp is removed in 0930972

the wrapping bug is fixed in e16a487 and covered with a new render test: https://github.com/mapbox/mapbox-gl-test-suite/commit/a7e5668c001b6f63f618604629bb6fbbca935682

do you see anything else or can this be squashed and merged (once tests ci passes)?

**Comment by jfirebaugh** [comment]
LGTM

**Pull request merged**